hyperdrive.go

```diff
@@ -113,3 +166,112 @@ func (hyperdrive *hyperdrive) AcceptShard(shard shard.Shard, blockchain block.Bl
 
 	r.Init()
 }
+
+// For a SignedBlock to be valid:
+// 1. must not be nil
+// 2. have valid blockTime, Round, and Height
+// 3. have a valid signature
+// 4. signatory belongs to the same shard
+// 5. parent header is the block at head of the shard's blockchain
+func (hyperdrive *hyperdrive) validateBlock(shardHash sig.Hash, signedBlock block.SignedBlock) bool {
+	if signedBlock.Block.Equal(block.Block{}) {
+		return false
+	}
+	if signedBlock.Time.After(time.Now()) {
+		return false
+	}
+	if signedBlock.Round < 0 || signedBlock.Height < 0 {
+		return false
+	}
+
+	// Verify the signatory of the signedBlock
+	signatory, err := hyperdrive.signer.Verify(signedBlock.Block.Header, signedBlock.Signature)
+	if err != nil {
+		return false
+	}
+	if !signatory.Equal(signedBlock.Signatory) {
+		return false
+	}
+	if !hyperdrive.isSignatoryInShard(shardHash, signatory) {
+		return false
+	}
+
+	// Is block.ParentHeader (at H) == block.Header (at H-1)?
+	if _, ok := hyperdrive.shardBlockchains[shardHash]; !ok {
+		return false
+	}
+	blockchain := hyperdrive.shardBlockchains[shardHash]
+	parent, ok := blockchain.Head()
+	if !ok {
+		return false
+	}
+
+	return parent.Header.Equal(signedBlock.ParentHeader)
+}
+
+// For a Polka to be valid:
+// 1. must not be nil
+// 2. have a non-negative Round and Height
+// 3. all the signatures inside the polka must be valid and
+//    belong to signatories within the same shard
+// 4. have a valid block (if block is not nil)
+//
+// validatePolka assumes that `polka.Signatures` are ordered to match
+// the order of `polka.Signatories`.
+func (hyperdrive *hyperdrive) validatePolka(shardHash sig.Hash, polka block.Polka) bool {
```

Comment:
We also need to check that the number of Signatures/Signatories are equal to each other, and greater than or equal to the consensus threshold of the Shard.

Reply:
👍 Fixed this.